### PR TITLE
Implement NV097_LAUNCH_TRANSFORM_PROGRAM

### LINKED ISSUE #309: Menu Stretching/Distortion and/or Glitches in Games
### Title Information

* https://xemu.app/titles/4950000a/#Baldur-s-Gate-Dark-Alliance-II
* https://xemu.app/titles/54540012/#Close-Combat-First-to-Fight-United-States-Marines
* https://xemu.app/titles/4541005a/#Need-for-Speed-Underground-2
* https://xemu.app/titles/53450036/#OutRun-2 (weird reorientation menu issue)
* https://xemu.app/titles/4541005f/
* https://xemu.app/titles/494f0005/#Metal-Slug-3 
* https://xemu.app/titles/454100ac/
* https://xemu.app/titles/41560022/
* https://xemu.app/titles/42550004/#Double-S-T-E-A-L-The-Second-Clash
* https://xemu.app/titles/42530009 (This one is rather minor, only in the bottom right area)
* https://xemu.app/titles/4e4d0019 (This one is rather severe)
* https://xemu.app/titles/5454007c (This one is a black void, nothing rendered other than some jumbled textures.)
* https://xemu.app/titles/53450088 (shoved up in the top left of the screen, for the whole game, but added because of the menu issue)
* https://xemu.app/titles/45410060 (It's workable, has some black lines and black bars around the menu area)
* https://xemu.app/titles/54540008 (Bad Stretching Issue)
* https://xemu.app/titles/4d530053 (workable, glitchy menu area however)
* https://xemu.app/titles/53450026 (glitches around every 5 seconds)
* https://xemu.app/titles/53538005 (background glitches and distorts)
* https://xemu.app/titles/56550036 (very glitchy)
* https://xemu.app/titles/4156004f (menu's have no rendered buttons, very glitchy)

### Bug Description

These games have menu stretching issues meaning that the game can't easily be accessed and/or obstructed because of these menu issues, by correcting this we can allow more games to have menu accessibility, thus why it is being documented.

### Additional Context

This is mostly for documentation purposes and I would love for anyone to include their games into this report, and I will add them into this report as I see them. Cataloging the titles with these issues allows us to further enhance xemu by working on the games that are close to starting or have minor issues preventing them from being playable.

### LINKED ISSUE #537: Some title have heavy exploding geometry ingame and menus
### Title

https://xemu.app/titles/5454001b/#Robotech-Invasion
https://xemu.app/titles/5454007c/#Spy-vs-Spy
https://xemu.app/titles/4e4d0019/#Curious-George

### Bug Description

When in menu and in-game server game breaking exploding geometry make these games unplayable.
Robotech: invasion
Screenshot from 2021-11-02 17-56-38
Others are similar in-game and menus are mostly a black screen but change when moving the menu items.
Steps to reproduce this issue:
1.Insert disk
2.Skip intro
3.menu broken

### Expected Behavior

menus should look like this.
Curious George:
Screenshot from 2021-11-07 19-44-50
Spy vs Spy:
Screenshot from 2021-11-07 19-48-42
Robotech Invasion:
Screenshot from 2021-11-07 19-49-42

### xemu Version

Version: 0.6.2
Branch: master
Commit: bbcd576
Date: Sat Oct 30 23:09:11 UTC 2021
Last-known Working: Unknown

### System Information

Field | Value
-- | --
OS | Ubuntu 20.04.3 LTS
CPU | Intel(R) Pentium(R) CPU 3825U @ 1.90GHz
Graphics Device | Mesa Intel(R) HD Graphics (BDW GT1)
Graphics Driver | Drivers: 4.6 (Core Profile) Mesa 21.0.3

### Additional Context

_No response_

### LINKED ISSUE #1072: Spy-vs-Spy: Main menu geometry is corrupt
### Title

https://xemu.app/titles/5454007c/#Spy-vs-Spy

### Bug Description

Note: This is an intentional dup of the generic #309 and #537 so I have a place to put my game-specific analysis.

The main menu in Spy Vs Spy renders with highly distorted geometry, most of the draws are clipped entirely due to the fact that the w component of the programmable vertex shader is very negative.

E.g., one of the screen space vertices: `{320.7315979004, 253.5145263672, 49013.19921875,  -100978.4765625}`

Screenshot_20220617_070622

### Expected Behavior

The menu should not be corrupted, and draws should generally be unclipped and projected properly.

### xemu Version

Happens in 0.7.39 and has apparently been happening for a long time (or forever), at least since 0.6.2

### System Information

CPU:          Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
OS Platform:  Linux
OS Version:   Ubuntu 21.10
Manufacturer: NVIDIA Corporation
GPU Model:    NVIDIA GeForce GTX 1070/PCIe/SSE2
Driver:       4.0.0 NVIDIA 470.129.06
Shader:       4.00 NVIDIA via Cg compiler

### Additional Context

The game uses S32K mode for the vertex positions, which is not something I've come across before. I wrote a [quick test](https://github.com/abaire/nxdk_vsh_tests/blob/main/src/tests/vertex_data_array_format_tests.cpp) to validate xemu's behavior and it seems to align with what I see in HW, so there's something else going on here.

### LINKED ISSUE #1074: nv2a: 0x1e80 - 0x1e90 are unhandled (NV097_SET_TRANSFORM_DATA and NV097_LAUNCH_TRANSFORM_PROGRAM)
### Bug Description

Thanks to @jackchentwkh for [the information](https://github.com/mborgerson/xemu/issues/1072#issuecomment-1159311417) in #1072: 

```
0x1e8x NV097_SET_TRANSFORM_DATA is supposed to be an array of four float values
that are assigned to the V0 register used in vertex shader.

0x1e90 NV097_LAUNCH_TRANSFORM_PROGRAM shall launch the vertex shader inside 
GPU, and the argument comes with it is the starting point (slot) of the vertex shader, which
shall be sitting within 0 to 135.
```

This is used in Spy vs Spy during the main menu rendering and may or may not be a contributing factor to #1072, so I'm splitting it into a new bug in case it turns out to be unrelated. 

### Expected Behavior

Handling of these commands should be implemented. More research will need to be done to determine if the target vertex shader can use any other inputs as well as to understand what is done with the outputs (it may just rely on the standard nv2a behavior of persisting register values until they are explicitly changed).

### xemu Version

Affects all xemu versions through 0.7.39

### System Information

n/a, this is not system specific

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by abaire**
Adds a CPU-based vertex shader emulator to handle vertex state shaders.

Fixes #1074 
Fixes #1072 
Fixes #309 
Fixes #537 

Improves #670, but it looks like there are still some font or scaling issues 

[Tests](https://github.com/abaire/nxdk_vsh_tests/blob/main/src/tests/cpu_shader_tests.cpp) - note that there are cases where the CPU shader results do not precisely match HW but are fairly close (typically when dealing with numbers at the extreme ranges of available precision).

Screen Shot 2022-06-19 at 15 07 47

### CODE REVIEW

**Comment by abaire** [comment]
I still need to run a lot of HW tests to validate my vsh emulator, but the current state is enough to fix Spy vs. Spy's loading screen and menu.

**Comment by ko81e24wy** [comment]
NBA JAM in game.
xemu-2022-06-20-17-13-44

**Comment by Triticum0** [comment]
~~NBA game has been improved in the menus but is a bit broken~~
xemu-2022-06-20-10-30-58
xemu-2022-06-20-10-31-18
xemu-2022-06-20-10-31-22
xemu-2022-06-20-10-31-24
Edit This has been fixed

**Comment by Triticum0** [comment]
Fixed Robotech Invasion. Now ingame
xemu-2022-06-20-10-36-14
xemu-2022-06-20-10-36-29

**Comment by Triticum0** [comment]
[Curious-George](https://xemu.app/titles/4e4d0019/#Curious-George) is fixed and now goes in-game
xemu-2022-06-20-10-42-42

**Comment by jackchentwkh** [comment]
@abaire I am not familiar with nv2a vsh code, but for me it seems to disasm vsh code from SVS doesn't fit with your implementation in detail. 
for the disasm MUL codes below

MUL R6.xyzw, R2, v0.x
MUL R7.xyzw, R3, v0.y
MUL R8.xyzw, R4, v0.z
MUL R9.xyzw, R5, v0.w

  it seems every components of R2 are all multiplied with the same float, v0.x, from the 1st line above. but in your implementation, each component of R2 shall be multiplied with corresponded component of v0. so which way is correct? or shall we need a overload for MUL operator here?

Similar for DP4 operation.

DP4 c[104].x, R2, c[92]
DP4 c[104].y, R3, c[92]
DP4 c[104].z, R4, c[92]
DP4 c[104].w, R5, c[92]

the DP4 result of R2,c[92] is stored to c[104].x, not the all 4 floats of c[104]. because later the other 3 lines of vsh codes obviously change the other 3 floats of c[104]. so again, which way is correct? or do we need a overload for DP4 here?

**Comment by abaire** [comment]
> @abaire I am not familiar with nv2a vsh code, but for me it seems to disasm vsh code from SVS doesn't fit with your implementation in detail. for the disasm MUL codes below
> 
> MUL R6.xyzw, R2, v0.x MUL R7.xyzw, R3, v0.y MUL R8.xyzw, R4, v0.z MUL R9.xyzw, R5, v0.w
> 
> it seems every components of R2 are all multiplied with the same float, v0.x, from the 1st line above. but in your implementation, each component of R2 shall be multiplied with corresponded component of v0. so which way is correct? or shall we need a overload for MUL operator here?
> 
> Similar for DP4 operation.
> 
> DP4 c[104].x, R2, c[92] DP4 c[104].y, R3, c[92] DP4 c[104].z, R4, c[92] DP4 c[104].w, R5, c[92]
> 
> the DP4 result of R2,c[92] is stored to c[104].x, not the all 4 floats of c[104]. because later the other 3 lines of vsh codes obviously change the other 3 floats of c[104]. so again, which way is correct? or do we need a overload for DP4 here?

As I responded in https://github.com/abaire/nv2a_vsh_cpu, the "cpu" methods intentionally do not handle swizzling or masking, they just dumbly apply the inputs to the output. Overloads should not be necessary for correct functionality, and going that path would substantially bloat the code. Swizzles are expanded such that ".x" => ".xxxx" and ".yz" => ".yzzz", this is normal nv2a behavior. Writemasks are handled in the emulator code after the operation completes such that only the components that are supposed to be assigned are used.
It's definitely possible that there are bugs (I expect it, which is why this PR is in draft until I can test on HW), but as you point out it has the correct behavior in games, so it follows that those bugs are more likely edge cases than fundamental errors like obliterating results with subsequent masked writes or using totally incorrect inputs.

Also, please don't post essentially the same comments across multiple places, it multiplies the time I have to take to respond.

**Comment by Zankkk** [comment]
When are we seeing this gets pulled/merged into master? really looking forward to it <3

**Comment by abaire** [comment]
> When are we seeing this gets pulled/merged into master? really looking forward to it <3

Same answer as on every PR: when it's done, and when it's reviewed, both of which take an unbounded amount of time. This is a volunteer project and both creating PRs and reviewing them take a large amount of effort. Please refrain from adding noise to PR comment streams; if you're not reporting that something is not working, you're absolutely welcome to look forward to things silently or comment in Discord (which, for clarity, still will not result in anything moving faster or slower, but at least will not clutter GitHub notifications).

**Comment by abaire** [comment]
I wrote some tests that compare HW results for the various ops to the CPU shader results and now have things pretty close for all operations. It feels like I'm at the point of diminishing returns wrt. the remaining differences so I think this is ready for review.

**Comment by Triticum0** [comment]
This is for reference as it only fixes some of  #309. Also, line out means it has been fixed or an open pr fixes the issue.
    ~~https://xemu.app/titles/4950000a/#Baldur-s-Gate-Dark-Alliance-II~~ #390
    ~~https://xemu.app/titles/54540012/#Close-Combat-First-to-Fight-United-States-Marines~~ #390
    ~~https://xemu.app/titles/4541005a/#Need-for-Speed-Underground-2~~ #677
    ~~https://xemu.app/titles/4541007b/#Need-for-Speed-Most-Wanted~~ #677
    ~~https://xemu.app/titles/53450036/#OutRun-2~~ #390 (weird reorientation menu issue)
    ~~https://xemu.app/titles/4541005f/The Lord of the Rings: The Third Age~~#735
    https://xemu.app/titles/494f0005/#Metal-Slug-3
    ~~https://xemu.app/titles/454100ac/Madden NFL 08~~ #390
    ~~https://xemu.app/titles/41560022/MTX Mototrax~~#677
   https://xemu.app/titles/42550004/#Double-S-T-E-A-L-The-Second-Clash
    ~~AMF Bowling 2004~~#390 (This one is rather minor, only in the bottom right area)
    ~~Curious George~~ #1080 (This one is rather severe)
    ~~Spy vs. Spy~~ #1080 (This one is a black void, nothing rendered other than some jumbled textures.)
    ~~OutRun 2006: Coast 2 Coast~~#390 (shoved up in the top left of the screen, for the whole game, but added because of the menu issue)
    ~~The Urbz: Sims in the City~~ (Need retesting) (It's workable, has some black lines and black bars around the menu area)
    ~~Midnight Club II~~#390 (Bad Stretching Issue)
    ~~Grabbed by the Ghoulies~~#962 (workable, glitchy menu area however)
    ESPN NHL Hockey (glitches around every 5 seconds)
    Outlaw Golf: Holiday Golf (Need retesting) (background glitches and distorts)
    Crash Twinsanity #319 improved the issue but still broken on mac (very glitchy)
    Total Club Manager 2005 (Need retesting) (minor glitches)
    ~~American Chopper 2: Full Throttle~~ #1045

File: hw/xbox/nv2a/pgraph.c [hunk 1]
```diff
@@ -3689,6 +3690,33 @@ DEF_METHOD(NV097, SET_SHADER_OTHER_STAGE_INPUT)
...
+}
+
+DEF_METHOD(NV097, LAUNCH_TRANSFORM_PROGRAM)
+{
+    int shader_slot = parameter;
+    Nv2aVshProgram program;
+
+    Nv2aVshParseResult result = nv2a_vsh_parse_program(
+            &program,
+            pg->program_data[shader_slot],
```
Comment (mborgerson) [inline hw/xbox/nv2a/pgraph.c]: Let's add a check to prevent OOB reads from `program_data` and possible underflow on max length expression below

Comment (abaire) [reply hw/xbox/nv2a/pgraph.c]: Added an assert

Comment (mborgerson) [reply hw/xbox/nv2a/pgraph.c]: Let's make shader_slot into `unsigned int program_start` and remove the >= 0 test

Comment (abaire) [reply hw/xbox/nv2a/pgraph.c]: Done

Comment (mborgerson) [reply hw/xbox/nv2a/pgraph.c]: Can you change the name to program_start too, we use slot elsewhere to refer to the token word

Comment (abaire) [reply hw/xbox/nv2a/pgraph.c]: Sure, done.

Comment (mborgerson) [reply hw/xbox/nv2a/pgraph.c]: Thanks

**Comment by mborgerson** [comment]
Nice work! Thanks

**Comment by Triticum0** [comment]
This also fixed metal slug 3
